### PR TITLE
External js error logging service

### PR DESCRIPTION
**Pull request by artpi**
### TLDR
This PR connects Calypso to the error-logging service 'Sentry': https://getsentry.com/welcome/
We have 14 days to test it for free.
This is more of a proof-of-concept, but I would love to see some kind of solution for this.

## Why?
Recently, `plans/compare` was broken for more than 2 weeks. This is kinda embarassing and we shouldn't have let this happen.
We have a lot of users that click through Calypso and test stuff for us. Even we do the same on Staging. Lets make use of it!

## Why service?
There are similar PRs that aim to do the same on our infrastructure- #1949 

### Pros of using a service
- Someone will be actually improving the UI for tracking errors
- This one has nice integrations (Slack, Github...)
- Having error handling outside the application is in my opinion safer, because it is less likely to fail at the same time as app

## I wanna see errors!
Please ping me, I will add you to the project.

CC
@drewblaisdell @scruffian @dmsnell @blowery

### CODE REVIEW

File: shared/lib/error-reporting/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	}
+}
+
+export function installErrorReporter( env ) {
+	return {
+		scriptUrl: 'https://cdn.ravenjs.com/2.0.1/raven.min.js',
+		installCode: `
+			if ( Raven ) {
+				Raven.config( 'https://5bae20cfadc949b48ffd4f96c1900b6c@app.getsentry.com/62683', {
+				  whitelistUrls: [/calypso\.localhost/, /wordpress\.com/],
```
Comment (artpi) [inline shared/lib/error-reporting/index.js]: Of course, I would restrict it only to `wordpress.com`, so it only works in production and staging

**Comment by rralian** [comment]
Worth looping in @bazza as an external service has security implications as the service (or anyone able to hack it) can then run arbitrary javascript on our site. It's also another vector for bugs as some bug in their code could then effectively break calypso.

Though I see that getsentry are actually [open source](https://github.com/getsentry) which is pretty damn cool. My two cents are that a minimal footprint client library from a company we trust is probably worth it and better than us building something that I don't think we're likely to focus on.

**Comment by artpi** [comment]
I chose specifically them because:
- Open Source
- You can install it 'On Premise' - then security is our concern
- Incredible integrations
- Other big brands seem to trust them

**Comment by artpi** [comment]
Alternatively, I can rewrite this code to gather JS errors in Google Analytics, where we already have user data etc.

**Comment by gwwar** [comment]
:+1: Nice to see us looking at error aggregators. They're very useful tools and are helpful in both squashing live bugs and sleuthing for support. There's also rollbar, airbrake and number of other competitors to consider.

An on prem contract may cost us :money_with_wings:  though.

**Comment by artpi** [comment]
I just want to get the ball rolling, this PR may be refactored to use any service, so we need this kind of discussion 
:)

**Comment by rralian** [comment]
> We have 14 days to test it for free.

@artpi Has the 14-day countdown already started?

I think the main question is whether we want to use a hosted service or host our own back-end for exception logging. I seem to recall some discussion about preferring to avoid external javascript services that we run on wordpress.com (I mean, of course), but I can't find any discussion threads about it. @bazza and/or @xknown can you chime in here?

I think the other question about using Sentry is pricing ([price list](https://getsentry.com/pricing/)). And honestly the pricing seems pretty reasonable for our needs. I don't know if we even need more than their first tier to get a sampling of errors. Maybe @apeatling or @nb could chime in here?

It seems like a no-brainer to try the 14-day trial here, just looking for some signoff to give it a try.

**Comment by rralian** [comment]
I briefly chatted with @bazza and can summarize that yes, he's not crazy about loading external javascript on wordpress.com. It gives these services arbitrary javascript execution, access to a user's cookies, and an error in their system can kill the whole page for a user. He also pointed out that we already had javascript exception logging working back in the newdash/Atlas days (company jargon for anyone following along), so we already have a back-end setup with some logic to accept js-errors, why not start by making use of that.

**Comment by artpi** [comment]
> Has the 14-day countdown already started?
Yes and it's coming to an end, but I'm sure we can generate a new key if we decide to.

> we already have a back-end setup with some logic to accept js-errors, why not start by making use of that.
Ok, so should I close the PR?

**Comment by artpi** [comment]
I am closing this PR.
What will happen is:
1. We will merge https://github.com/Automattic/wp-calypso/pull/1949/files
2. We'll see how the errors go
3. We will install sentry on our own servers
